brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2018 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "brave/utility/tor/tor_impl.h"
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/process/kill.h"
+#include "base/process/launch.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task_scheduler/post_task.h"
+#include "base/threading/thread.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signal) {
+  int error = errno;
+  char ch = 0;
+  write(pipehack[1], &ch, 1);
+  errno = error;
+}
+
+static void SetupPipeHack() {
+  if (pipe(pipehack) == -1)
+    LOG(ERROR) << "pipehack";
+
+  int flags;
+  for (size_t i = 0; i < 2; ++i) {
+    if ((flags = fcntl(pipehack[i], F_GETFL)) == -1)
+      LOG(ERROR) << "get flags";
+    if (i == 1)
+      flags |= O_NONBLOCK;
```

Comment:
Can you add a brief comment explaining this?  Something like:

(a) we don't want to block the write end, which is the signal handler `SIGCHLDHandler`, since there's always space for PIPE_BUF bytes, and all we need is one byte to notify the thread; but
(b) we do want to block the read end, which is the thread `MonitorChild` that uses reads from the pipe to wait for the SIGCHLD signals to arrive.

(If Unix weren't stupid, then we could just use `write_nonblock` in `SIGCHLDHandler` and `read_block` in `MonitorChild`, but there are no such distinct syscalls, so we have to rely on nonlocal state like this.)

Reply:
fixed in 4e72ec6